src/Compilers/Core/Portable/SourceGeneration/Nodes/BatchNode.cs

```diff
@@ -50,36 +53,101 @@ public NodeStateTable<ImmutableArray<TInput>> UpdateStateTable(DriverStateTable.
+                {
+                    newTable.AddEntry(sourceValues, EntryState.Added, stopwatch.Elapsed, sourceInputs, EntryState.Added);
+                }
+            }
+
+            return newTable.ToImmutableAndFree();
+        }
+
+        private (ImmutableArray<TInput>, ImmutableArray<(IncrementalGeneratorRunStep InputStep, int OutputIndex)>) GetValuesAndInputs(
+            NodeStateTable<TInput> sourceTable,
+            NodeStateTable<ImmutableArray<TInput>> previousTable,
+            NodeStateTable<ImmutableArray<TInput>>.Builder newTable)
+        {
+            // Do an initial pass to both get the steps, and determine how many entries we'll have.
             var sourceInputsBuilder = newTable.TrackIncrementalSteps ? ArrayBuilder<(IncrementalGeneratorRunStep InputStep, int OutputIndex)>.GetInstance() : null;
 
+            var entryCount = 0;
             foreach (var entry in sourceTable)
             {
-                // At this point, we can remove any 'Removed' items and ensure they're not in our list of states.
-                if (entry.State != EntryState.Removed)
-                    sourceValuesBuilder.Add(entry.Item);
-
-                // However, regardless of if the entry was removed or not, we still keep track of its step information
-                // so we can accurately report how long it took and what actually happened (for testing validation).
+                // Always keep track of its step information, regardless of if the entry was removed or not, so we
+                // can accurately report how long it took and what actually happened (for testing validation).
                 sourceInputsBuilder?.Add((entry.Step!, entry.OutputIndex));
+
+                if (entry.State != EntryState.Removed)
+                    entryCount++;
             }
 
-            var sourceValues = sourceValuesBuilder.ToImmutableAndFree();
-            var sourceInputs = newTable.TrackIncrementalSteps ? sourceInputsBuilder!.ToImmutableAndFree() : default;
+            var sourceInputs = sourceInputsBuilder != null ? sourceInputsBuilder.ToImmutableAndFree() : default;
 
-            if (previousTable.IsEmpty)
+            // First, see if we can reuse the entries from the previous table.
+            var result = tryReusePreviousTableValues(entryCount);
+            if (result.IsDefault)
             {
-                newTable.AddEntry(sourceValues, EntryState.Added, stopwatch.Elapsed, sourceInputs, EntryState.Added);
+                // If not, produce the actual values we need from sourceTable.
+                result = computeCurrentTableValues(entryCount);
             }
-            else if (!sourceTable.IsCached || !newTable.TryUseCachedEntries(stopwatch.Elapsed, sourceInputs))
+
+            return (result, sourceInputs);
+
+            ImmutableArray<TInput> tryReusePreviousTableValues(int entryCount)
```

Comment:
this does non-allocating pass looking to see if we can reuse the prior array.  it checks the lengths are the same then checks all non-removed items against the prior array.  if a match, we can reuse.